### PR TITLE
Added custom config path

### PR DESCRIPTION
**Pull request by muuvmuuv**
It is now possible to add a custom config path by setting `unibeautify.defaultConfig` to a specific path like: `/Users/marvinheilemann/Dropbox/Configs/.unibeautifyrc.json` and if someone wants to have the `defaultConfig` overridden by the workspace config, it can set `unibeautify.extendConfig` to `true`

### CODE REVIEW

**Comment by muuvmuuv** [comment]
@Glavin001 check my latest commit. I think it is ready to merge.

File: package.json [hunk 1]
```diff
@@ -1,7 +1,7 @@
 {
     "name": "unibeautify-vscode",
     "displayName": "Unibeautify - Universal Formatter",
-    "version": "0.5.0",
+    "version": "0.5.1",
```
Comment (szeck87) [inline package.json]: Please don't change the version in PR's

File: package.json [hunk 2]
```diff
@@ -94,7 +88,7 @@
         "diff": "^3.3.0",
         "eslint": "^4.19.1",
         "js-beautify": "^1.7.5",
-        "prettier": "^1.12.1",
+        "prettier": "^1.13.5",
```
Comment (szeck87) [inline package.json]: Please leave dependencies unrelated to this feature as they are.  We're working on updating them and keeping them up to date via Renovate.

File: package.json [hunk 3]
```diff
@@ -107,8 +101,18 @@
                     "type": "boolean",
                     "default": true,
                     "description": "Enable/disable Unibeautify formatter"
+                },
```
Comment (szeck87) [inline package.json]: Please rebase with the current master branch and apply these changes on top of it

File: package.json (hunk 1 above)
Comment (Glavin001) [reply package.json]: Please revert.

File: package.json [hunk 4]
```diff
@@ -10,9 +10,7 @@
         "color": "#ffe04b",
         "theme": "light"
     },
-    "activationEvents": [
-        "*"
-    ],
+    "activationEvents": ["*"],
```
Comment (Glavin001) [inline package.json]: Please no formatting changes to `package.json`, as `npm` uses `JSON.stringify` and this will cause changes back after running `npm install --save`.

Comment (Glavin001) [reply package.json]: @szeck87 We really need to create `JSON.stringify` beautifier to handle these files. So much to do 😫

File: package.json [hunk 5]
```diff
@@ -107,8 +101,18 @@
                     "type": "boolean",
                     "default": true,
                     "description": "Enable/disable Unibeautify formatter"
+                },
+                "unibeautify.defaultConfig": {
+                    "type": "string",
+                    "default": null,
+                    "description": "Type in a path to your .unibeautifyrc outside the workspace folder. This will override the projects .unibeautifyrc config file. Note: The file path must be absolute for a directory or file!"
```
Comment (Glavin001) [inline package.json]: Consider rephrasing:

```diff
- "description": "Type in a path to your .unibeautifyrc outside the workspace folder. This will override the projects .unibeautifyrc config file. Note: The file path must be absolute for a directory or file!"
+ "description": "Absolute path to your .unibeautifyrc file outside the workspace folder. This will override the workspace's own .unibeautifyrc config file."
```

File: src/EditProvider.ts [hunk 6]
```diff
@@ -99,16 +100,84 @@ export class EditProvider
...
+    path: string = vscode.workspace.rootPath
   ): Promise<LanguageOptionValues> {
     try {
-      const explorer = cosmiconfig("unibeautify", {});
+      /**
+       * Use a different `.unibeautify` config when vscode config is set and
+       * define if the project config should extend the vscode config.
+       *
+       * @argument {unibeautify.defaultConfig} string path to a file
+       * @argument {unibeautify.extendConfig} boolean extend or not
```
Comment (Glavin001) [inline src/EditProvider.ts]: A comment should be unnecessary in this case so long as we have well written / readable code. Please remove 😄.

Now to look below and make sure we have readable code 😜 .

File: src/index.ts [hunk 7]
```diff
@@ -11,7 +11,9 @@ export function activate(context: vscode.ExtensionContext) {
   }
   unibeautify.loadBeautifiers(beautifiers);
   const { supportedLanguages } = unibeautify;
-  return EditProvider.beautifyOptions().then(options => {
+  return EditProvider.beautifyOptions().then((options: any) => {
```
Comment (Glavin001) [inline src/index.ts]:
```diff
- options: any
+ options: LanguageOptionValues
```

File: src/index.ts [hunk 8]
```diff
@@ -11,7 +11,9 @@ export function activate(context: vscode.ExtensionContext) {
...
+    if (!options) return console.error("No options:", options);
```
Comment (Glavin001) [inline src/index.ts]: What is the use case for this? How was it not working before?

File: src/EditProvider.ts [hunk 9]
```diff
@@ -99,16 +100,84 @@ export class EditProvider
...
+       */
+      const vscodeSettings = vscode.workspace.getConfiguration("unibeautify");
```
Comment (Glavin001) [inline src/EditProvider.ts]: Use the `UnibeautifyVSCodeSettings` interface from https://github.com/Unibeautify/vscode/blob/master/src/index.ts#L65-L67

I think you will need to move `UnibeautifyVSCodeSettings` from `index.ts` to this file then import into `index.ts`.

```diff
- const vscodeSettings = vscode.workspace.getConfiguration("unibeautify");
+ const vscodeSettings: UnibeautifyVSCodeSettings = vscode.workspace.getConfiguration("unibeautify");
```

File: src/EditProvider.ts [hunk 10]
```diff
@@ -99,16 +100,84 @@ export class EditProvider
...
+      let defaultConfigOverride = vscodeSettings.defaultConfig;
+      let extendConfig = vscodeSettings.extendConfig;
+      let workspace = path;
+
+      const explorerOptions: any = {};
```
Comment (Glavin001) [inline src/EditProvider.ts]: There should be any interface from `cosmiconfig` types to use instead of `any`.

File: src/EditProvider.ts [hunk 11]
```diff
@@ -58,7 +59,7 @@ export class EditProvider
     const fileExtension = this.fileExtensionForDocument(document);
     const filePath = document.fileName;
     const projectPath = vscode.workspace.rootPath;
-    return EditProvider.beautifyOptions(filePath).then(beautifyOptions => {
+    return EditProvider.beautifyOptions(projectPath).then(beautifyOptions => {
```
Comment (Glavin001) [inline src/EditProvider.ts]: What is the reason for this change? Previously this allowed Unibeautify to lookup the configuration file closest to the `filePath`.

File: src/EditProvider.ts [hunk 12]
```diff
@@ -99,16 +100,84 @@ export class EditProvider
   }
 
   public static beautifyOptions(
-    filePath: string = vscode.workspace.rootPath
+    path: string = vscode.workspace.rootPath
```
Comment (Glavin001) [inline src/EditProvider.ts]: Is this a file path, directory path, or both?

File: src/EditProvider.ts [hunk 13]
```diff
@@ -99,16 +100,84 @@ export class EditProvider
...
+       * @argument {unibeautify.extendConfig} boolean extend or not
+       */
+      const vscodeSettings = vscode.workspace.getConfiguration("unibeautify");
+      let defaultConfigOverride = vscodeSettings.defaultConfig;
+      let extendConfig = vscodeSettings.extendConfig;
+      let workspace = path;
+
+      const explorerOptions: any = {};
+      if (defaultConfigOverride) {
+        path = defaultConfigOverride;
```
Comment (Glavin001) [inline src/EditProvider.ts]: The variable `path` is passed as an argument into this function. It is bad to reassign its value.

See these linting rules:
- https://eslint.org/docs/rules/no-param-reassign
- https://palantir.github.io/tslint/rules/no-parameter-reassignment/

We should add the linting rule to our Travis CI checks, if it is not already. cc @szeck87

File: src/EditProvider.ts [hunk 14]
```diff
@@ -99,16 +100,84 @@ export class EditProvider
...
+        return vscode.window.showErrorMessage(
+          `We could not find your configuration file: ${path}. Please correct your path, otherwise the plugin will not work!`
+        );
+      }
     } catch (error) {
       return Promise.reject(error);
     }
   }
+
+  public static extend(...obj: any[]) {
```
Comment (Glavin001) [inline src/EditProvider.ts]: This is going to need unit tests, if it does not already have some.

I would prefer we use https://www.npmjs.com/package/lodash.merge since it is tested and used by many already.

File: src/EditProvider.ts [hunk 15]
```diff
@@ -99,16 +100,84 @@ export class EditProvider
...
+        path = defaultConfigOverride;
+      }
+
+      const explorer = cosmiconfig("unibeautify", explorerOptions);
       const defaultConfig: LanguageOptionValues = {};
-      return explorer
-        .search(filePath)
-        .then(result => (result ? result.config : defaultConfig));
+
+      let isDirectory = fs.lstatSync(path).isDirectory();
```
Comment (Glavin001) [inline src/EditProvider.ts]: @muuvmuuv I don't think I want `unibeautify.defaultConfig` to support being a path to a directory. I would prefer it only support file path to `.unibeautifyrc` file and use `explorer.load(filePath)`.

cc @szeck87 thoughts?

File: src/EditProvider.ts [hunk 16]
```diff
@@ -99,16 +100,84 @@ export class EditProvider
...
+          .then(returnResult)
+          .catch(catchError);
+      } else {
+        return explorer
+          .search(path)
+          .then(returnResult)
+          .catch(catchError);
+      }
+
+      function returnResult(result: any) {
```
Comment (Glavin001) [inline src/EditProvider.ts]: The `beautifyOptions` function is growing too large. It should either be broken down into multiple methods or perhaps some of the code can be deleted and/or simplified. Such as by removing support for `unibeautify.defaultConfig` to be a directory.

File: package.json [hunk 17]
```diff
@@ -107,8 +101,18 @@
...
+                },
+                "unibeautify.extendConfig": {
```
Comment (Glavin001) [inline package.json]: Part of me feels like supporting `extendConfig` is something which can come later. I support allowing users to set `defaultConfig` such that they can have a file always to fallback on for beautification settings.
However, once we start incorporating merging of multiple `.unibeautifyrc` files, it's another thing to worry about and I'm not comfortable merging yet.

Features are easy to add and hard to remove 😜 .

@muuvmuuv could you remove `extendConfig` from this PR for now? I'd like to get `defaultConfig` support merged in ASAP and I want to keep this Pull Request laser focused. Thanks.

cc @szeck87 thoughts?

File: src/EditProvider.ts (hunk 11 above)
Comment (muuvmuuv) [reply src/EditProvider.ts]: because in version 5 of `cosmiconfig` it will `search` inside a directory or `load` a config file. I tried `filePath`but got an error every time, because `cosmiconfig` was trying to load the changed file as a config. 

```
notificationsAlerts.ts:42 We could not find your configuration file: 
/Users/marvinheilemann/Documents/Projects/_extensions/unibeautify-vscode/testProject/typescript/test.ts. 
Please correct your path, otherwise the plugin will not work!
```
error from line 159

Anyway we could leave it blank, because this is the default passed string on line 103

File: src/EditProvider.ts (hunk 12 above)
Comment (muuvmuuv) [reply src/EditProvider.ts]: It is both, depending on what a call is passing to the function ()

File: src/EditProvider.ts (hunk 9 above)
Comment (muuvmuuv) [reply src/EditProvider.ts]: How do you mean this, I did this now:

```
import { UnibeautifyVSCodeSettings } from "./index";
.......
      const vscodeSettings: UnibeautifyVSCodeSettings = vscode.workspace.getConfiguration(
        "unibeautify"
      );
      let defaultConfigOverride = vscodeSettings.defaultConfig;
```

ang got this tslint error:
```
[ts]
Der Typ "WorkspaceConfiguration" kann dem Typ "UnibeautifyVSCodeSettings" nicht zugewiesen werden.
  Die Eigenschaft "enabled" fehlt im Typ "WorkspaceConfiguration".
const vscodeSettings: UnibeautifyVSCodeSettings
```

File: src/EditProvider.ts (hunk 15 above)
Comment (muuvmuuv) [reply src/EditProvider.ts]: I think both are good because you can choose your project root and it will search by its own for a config file. I‘ll exclude it in the next pr for now

File: src/EditProvider.ts (hunk 14 above)
Comment (muuvmuuv) [reply src/EditProvider.ts]: Yeah I‘ll add this when we decided to use `extendConfig`, for now I‘ll remove it completely

File: src/EditProvider.ts (hunk 16 above)
Comment (muuvmuuv) [reply src/EditProvider.ts]: I tried to shorten it to this:

```
      return explorer
        .load(pathToFileOrDirectory)
        .then(result => (result ? result.config : defaultConfig)
        .catch(() => vscode.window.showErrorMessage(`We could not find your configuration file: ${path}. Please correct your path, otherwise the plugin will not work!`));
```

but I‘m getting this error:

```
[ts]
Der Typ "Promise<string | LanguageOptionValues>" kann dem Typ "Promise<LanguageOptionValues>" nicht zugewiesen werden.
  Der Typ "string | LanguageOptionValues" kann dem Typ "LanguageOptionValues" nicht zugewiesen werden.
    Der Typ "string" kann dem Typ "LanguageOptionValues" nicht zugewiesen werden.
```

File: src/EditProvider.ts (hunk 15 above)
Comment (Glavin001) [reply src/EditProvider.ts]: Searching from project root should definitely be supported, as it is now.

However, the `unibeautify.defaultConfig` option should be a file and not a directory, as this should only be used for non-standard and out-of-project-root configuration file.

File: src/EditProvider.ts (hunk 16 above)
Comment (Glavin001) [reply src/EditProvider.ts]: I do not speak that language so I cannot help you with analyzing the error message 😉 .
Take a look at https://github.com/Microsoft/TypeScript/wiki/What's-new-in-TypeScript#localized-diagnostics-on-the-command-line 😄

Comment (muuvmuuv) [reply src/EditProvider.ts]: @Glavin001 oh 😄 sorry, anyway I fixed it. btw it is german

**Comment by muuvmuuv** [comment]
@Glavin001 @szeck87 can u look over my latest pr :)

File: src/EditProvider.ts [hunk 18]
```diff
@@ -105,14 +107,42 @@ export class EditProvider
...
+        if (!result) return defaultConfig;
+
+        return result.config;
+      }
+
+      function catchError() {
+        return vscode.window.showErrorMessage(
+          `We could not find your configuration file: ${pathToConfig}. Please correct your path, otherwise the plugin will not work!`
+        );
```
Comment (Glavin001) [inline src/EditProvider.ts]: I'd prefer removing `pathToConfig` entirely and instead passing the various paths as an argument to generate a custom error message.

Consider the following:

```typescript
if (defaultConfigOverride) {
        return explorer
          .load(defaultConfigOverride)
          .then(returnResult)
          .catch(catchError(defaultConfigOverride));
}

return explorer
        .search(filePath)
        .then(returnResult)
        .catch(catchError(filePath));

function catchError(configPath) {
        return () => vscode.window.showErrorMessage(
          `We could not find your configuration file: ${configPath}. Please correct your path, otherwise the plugin will not work!`
        );
}
```

Comment (Glavin001) [reply src/EditProvider.ts]: On second glance, I see the `catchError` message says `configuration file` whereas the second call is actually searching for config file (i.e. `explorer.search(filePath)` ) not loading the file (i.e. `explorer.load(defaultconfigOverride)`) so the error message does not make sense in both cases.

I think something like this is more appropriate:

```typescript
if (defaultConfigOverride) {
        return explorer
          .load(defaultConfigOverride)
          .then(returnResult)
          .catch((error) => {
            vscode.window.showErrorMessage(`We could not find your configuration file: ${defaultConfigOverride}. Please correct your path, otherwise the plugin will not work!`);
             throw error;
        });
}

return explorer
        .search(filePath)
        .then(returnResult);
```

File: src/index.ts [hunk 19]
```diff
@@ -11,45 +11,47 @@ export function activate(context: vscode.ExtensionContext) {
...
-      ),
-      vscode.languages.registerDocumentRangeFormattingEditProvider(
-        documentSelector,
-        editProvider
-      )
-    );
-  });
+  return EditProvider.beautifyOptions().then(
+    (options: LanguageOptionValues) => {
+      console.log("Supported languages", supportedLanguages);
```
Comment (Glavin001) [inline src/index.ts]: This is a big change. I think it is only formatting changes?

**Comment by Glavin001** [comment]
@muuvmuuv @szeck87 Take a look at https://github.com/muuvmuuv/vscode/pull/1 . I think this fixes up most of my comments and then some 😄 .

File: src/index.ts (hunk 19 above)
Comment (muuvmuuv) [reply src/index.ts]: Yep because of the max-line-length it‘s wrapped.

**Comment by muuvmuuv** [comment]
@Glavin001 I see you already made the changes?

File: package.json [hunk 20]
```diff
@@ -110,6 +110,11 @@
                     "type": "boolean",
                     "default": true,
                     "description": "Enable/disable Unibeautify formatter"
+                },
+                "unibeautify.defaultConfig": {
+                    "type": "string",
+                    "default": null,
+                    "description": "Absolute path to your .unibeautifyrc file outside the workspace folder. This will override the workspace's own .unibeautifyrc config file."
```
Comment (Glavin001) [inline package.json]: > This will override the workspace's own .unibeautifyrc config file.

Will it? Or is this only the default *when `.unibeautifyrc` config file **does not exist within the workspace**?*

Looking at https://github.com/muuvmuuv/vscode/blob/a0c792544f8df0db4853ac9008d87f719b2536bf/src/EditProvider.ts#L121-L134 this description is accurate. However, I think this is not actually what we want.

@muuvmuuv I believe your use case was you did *not* have a `.unibeautifyrc` file within the workspace. You wanted to fallback on a `defaultConfig` in those cases. By adding `defaultConfig` for VSCode this allows you to configure within your `User Settings` of VSCode.

However, with the currently implementation, if you set `defaultConfig` in your `User Settings` (instead of `Workspace Settings`) then it would actually ignore the `.unibeautifyrc` within the workspace! This is not what we want.

cc @szeck87  right?

@muuvmuuv Could you swap around the login at https://github.com/muuvmuuv/vscode/blob/a0c792544f8df0db4853ac9008d87f719b2536bf/src/EditProvider.ts#L121-L134 such that the `defaultConfig` is a *fallback for when `.unibeautifyrc` is not found within the workspace? Thank you!

File: src/EditProvider.ts [hunk 21]
```diff
@@ -105,14 +109,36 @@ export class EditProvider
...
+          .catch(error => {
+            vscode.window.showErrorMessage(
+              `We could not find your configuration file: ${defaultConfigPath}.` +
+                "Please correct your path, otherwise the plugin will not work!"
+            );
+            throw error;
+          });
+      }
+
+      return explorer.search(searchStartPath).then(returnResult);
```
Comment (Glavin001) [inline src/EditProvider.ts]: See my comment above. I think `defaultConfig` should be the fallback, if `searchStartPath` (file path or workspace) does not have a `.unibeautifyrc` nearby.

File: test/jest-test-runner.ts [hunk 22]
```diff
@@ -6,7 +6,7 @@ import * as path from "path";
 import * as sourceMapSupport from "source-map-support";
 
 const testDirectory = path.resolve(__dirname, "../../test");
-const fromConfigDir = (filename: string) => path.resolve(__dirname, filename);
+// const fromConfigDir = (filename: string) => path.resolve(__dirname, filename);
```
Comment (Glavin001) [inline test/jest-test-runner.ts]: ?

File: testProject/.vscode/settings.json [hunk 23]
```diff
@@ -13,5 +13,7 @@
     ],
     "files.associations": {
         "*.myphp": "php"
-    }
-}
\ No newline at end of file
+    },
+    "unibeautify.defaultConfig": null,
+    "unibeautify.extendConfig": true
```
Comment (Glavin001) [inline testProject/.vscode/settings.json]: Remove `extendConfig`

File: package.json (hunk 20 above)
Comment (muuvmuuv) [reply package.json]: makes sense. I‘ll change this.

File: test/jest-test-runner.ts (hunk 22 above)
Comment (muuvmuuv) [reply test/jest-test-runner.ts]: This was by accident

**Comment by muuvmuuv** [comment]
@Glavin001 check them out `8c10204`

File: src/EditProvider.ts (hunk 9 above)
Comment (muuvmuuv) [reply src/EditProvider.ts]: Ahh sorry. I always forget that. I should turn my computer language to english!

The problem here ist, that on the next line, it does not find the `.defaultConfig` prop. I changed it a little bit to get it working. Are you fine with that:

```typescript
      const vscodeSettings: UnibeautifyVSCodeSettings = <any>vscode.workspace.getConfiguration(
        "unibeautify"
      );
      const defaultConfigFile = (vscodeSettings as any).defaultConfig;
```

File: package.json (hunk 20 above)
[the remark this answers is not in the document]
Comment (Glavin001) [reply package.json]: Sounds good! 👍

**Comment by muuvmuuv** [comment]
Hopefully the last changes 😁 @Glavin001 (do I have to mention you every time or do you receive a notification when I push something?)

File: src/EditProvider.ts [hunk 24]
```diff
@@ -112,8 +113,10 @@ export class EditProvider
     searchStartPath: string | undefined = vscode.workspace.rootPath
   ): Promise<LanguageOptionValues> {
     try {
-      const vscodeSettings = vscode.workspace.getConfiguration("unibeautify");
-      const defaultConfigFile = vscodeSettings.defaultConfig;
+      const vscodeSettings: UnibeautifyVSCodeSettings = <any>vscode.workspace.getConfiguration(
+        "unibeautify"
+      );
+      const defaultConfigFile = (vscodeSettings as any).defaultConfig;
```
Comment (Glavin001) [inline src/EditProvider.ts]: Should not need the casting to `any`:

```diff
-      const defaultConfigFile = (vscodeSettings as any).defaultConfig;
+      const defaultConfigFile = vscodeSettings.defaultConfig;
```

Add `defaultConfig` to interface at https://github.com/muuvmuuv/vscode/blob/f74329666cbc0b647afdfccf2f100a52e92a4576/src/index.ts#L67

Comment (muuvmuuv) [reply src/EditProvider.ts]: Ah okay :)

File: src/index.ts [hunk 25]
```diff
@@ -65,5 +65,6 @@ function vscodeSettings(): UnibeautifyVSCodeSettings {
 }
 
 export interface UnibeautifyVSCodeSettings {
+  defaultConfig: string;
```
Comment (Glavin001) [inline src/index.ts]: `UnibeautifyVSCodeSettings` should match `contributes.configuration` from `package.json`: https://github.com/muuvmuuv/vscode/blob/11c22960694b22f7b73e936cc36d83ab1bd49b40/package.json#L104-L122

For example, `enable` property: https://github.com/muuvmuuv/vscode/blob/11c22960694b22f7b73e936cc36d83ab1bd49b40/package.json#L110-L111

```
                "unibeautify.enabled": {
                    "type": "boolean",
                    "default": true,
```

So `enable` is of type `boolean | true` which can be simplified to `boolean`.

Let's look at `defaultConfig`: https://github.com/muuvmuuv/vscode/blob/11c22960694b22f7b73e936cc36d83ab1bd49b40/package.json#L115-L116

```
                "unibeautify.defaultConfig": {
                    "type": "string",
                    "default": null,
```

So `defaultConfig` is of type `string | null` not `string` (as it is in the interface at line 68).

It should look more like:

```
 export interface UnibeautifyVSCodeSettings {
-  defaultConfig: string;
+  defaultConfig: string | null;
```

😃 .

Also, I think `default: null` should actually be `default: undefined`. Since `undefined` is the value when it is not set, not `null`, if I am not mistaken?

@muuvmuuv please let me know if you have any questions. This PR is really close, there always seems to be *one more thing* 😝 .

Comment (muuvmuuv) [reply src/index.ts]: Okay, sorry I‘m not so good with typescript. Still learning, thanks!

Comment (muuvmuuv) [reply src/index.ts]: in package.json is only null valid, I have to wrap undefined in quotes to get it working. Are you ok with that?
Also I changed `string` to `string | undefined` inside the interface

Comment (Glavin001) [reply src/index.ts]: > I have to wrap undefined in quotes to get it working

Good to know. Then let's use `null`. `"undefined"` (string) is not what I want, if we can't have `undefined` then `null` will do 👍 . Use `defaultConfig: string | null`.

> Okay, sorry I‘m not so good with typescript. Still learning, thanks!

No worries! You're doing great! There's a lot going on here, with VSCode Extension's meta data in `package.json`'s `contributes.` and then keeping in sync with TypeScript 🎉 .
